gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/account/AccountEndpointHandler.java

```diff
@@ -122,6 +153,153 @@ public void updateProfile(RoutingContext routingContext) {
+                .doOnError(er -> AccountResponseHandler.handleUpdateUserResponse(routingContext, er.getMessage()))
+                .subscribe(res -> { /* no op */});
+    }
+
+    private EnrolledFactor buildEnrolledFactor(Factor factor, RoutingContext routingContext, User user, FactorStatus factorStatus) {
+        EnrolledFactor enrolledFactor = new EnrolledFactor();
+        enrolledFactor.setFactorId(factor.getId());
+        enrolledFactor.setStatus(factorStatus);
+        FactorTypeEnum factorType = FactorTypeEnum.getFactorTypeFromString(factor.getFactorType());
+        switch (factorType) {
+            case OTP:
+                enrolledFactor.setSecurity(new EnrolledFactorSecurity(SHARED_SECRET, SharedSecret.generate()));
+                break;
+            case SMS:
+                enrolledFactor.setChannel(new EnrolledFactorChannel(EnrolledFactorChannel.Type.SMS,
+                        routingContext.session().get(ConstantKeys.ENROLLED_FACTOR_PHONE_NUMBER)));
+                break;
+            case EMAIL:
+                Map<String, Object> additionalData = new Maps.MapBuilder(new HashMap())
+                        .put(FactorDataKeys.KEY_MOVING_FACTOR, generateInitialMovingFactor(user))
+                        .build();
+                // For email even if the endUser will contains all relevant information, we extract only the Expiration Date of the code.
+                // this is done only to enforce the other parameter (shared secret and initialMovingFactor)
+                getEnrolledFactor(factor, user).ifPresent(ef -> {
+                    additionalData.put(FactorDataKeys.KEY_EXPIRE_AT, ef.getSecurity().getData(FactorDataKeys.KEY_EXPIRE_AT, Long.class));
+                });
+                enrolledFactor.setSecurity(new EnrolledFactorSecurity(SHARED_SECRET,
+                        routingContext.session().get(ConstantKeys.ENROLLED_FACTOR_SECURITY_VALUE_KEY),
+                        additionalData));
+                enrolledFactor.setChannel(new EnrolledFactorChannel(EnrolledFactorChannel.Type.EMAIL,
+                        routingContext.session().get(ConstantKeys.ENROLLED_FACTOR_EMAIL_ADDRESS)));
+                break;
+            default:
+                throw new IllegalStateException("Unexpected value: " + factorType);
+        }
+        enrolledFactor.setCreatedAt(new Date());
+        enrolledFactor.setUpdatedAt(enrolledFactor.getCreatedAt());
+        return enrolledFactor;
+    }
+
+    private int generateInitialMovingFactor(User endUser) {
+        try {
+            SecureRandom secureRandom = SecureRandom.getInstance("SHA1PRNG");
+            secureRandom.setSeed(endUser.getUsername().getBytes(StandardCharsets.UTF_8));
+            return secureRandom.nextInt(1000) + 1;
+        } catch (NoSuchAlgorithmException e) {
+            return 0;
+        }
+    }
+
+    private Optional<EnrolledFactor> getEnrolledFactor(Factor factor, User endUser) {
+        return endUser.getFactors()
```

Comment:
this is the same implementation that's in MFAChallengeEndpoint

Reply:
Yes it was but since you started this feature, the code base has changed:  
https://github.com/gravitee-io/gravitee-access-management/blob/master/gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java#L304

Reply:
thank you resolved!